### PR TITLE
make use of native IntersectionObserver in 3p iframe

### LINKED ISSUE #5603: Stop using IntersectionObserver polyfill in ads if native support is present
Now that stable Chrome has IntersectionObserver we should use it instead of the postMessage mechanism.

CC @lannka

### PR DESCRIPTION
**Pull request by zhouyx**
close #5603

### CODE REVIEW

File: 3p/integration.js [hunk 1]
```diff
@@ -422,6 +423,16 @@ function triggerRenderStart(opt_data) {
  */
 function observeIntersection(observerCallback) {
   // Send request to received records.
+  if (window.IntersectionObserver) {
+    const io = new window.IntersectionObserver(changes => {
+      observerCallback(changes);
+    }, {
+      threshold: DEFAULT_THRESHOLD,
+    });
+    io.observe(window.document.documentElement);
```
Comment (zhouyx) [inline 3p/integration.js]: `window.document.documentElement` is the best I can do here. If it does not fill up our 3p iframe, we get different intersection changeEntry between native IO and polyfill IO.

File: 3p/integration.js [hunk 2]
```diff
@@ -422,6 +431,17 @@ function triggerRenderStart(opt_data) {
...
+  if (window.IntersectionObserver && window.IntersectionObserver.prototype) {
+    // use native IntersectionObserver if exist
+    const io = new window.IntersectionObserver(changes => {
+      observerCallback(changes);
+    }, {
+      threshold: DEFAULT_THRESHOLD,
+    });
+    io.observe(window.document.documentElement);
+    const unlistener = () => io.unobserve(window.document.documentElement);
+    return unlistener;
```
Comment (lannka) [inline 3p/integration.js]: return in the above line?

File: ads/_ping_.js [hunk 3]
```diff
@@ -22,7 +22,11 @@ import {dev} from '../src/log';
  */
 export function _ping_(global, data) {
   global.document.getElementById('c').textContent = data.ping;
-
+  if (!data.nativeIo) {
```
Comment (lannka) [inline ads/_ping_.js]: to be explicit, data.nativeIntersectionObserver

File: test/manual/fakead3p.amp.html [hunk 4]
```diff
@@ -48,6 +49,8 @@ <h2 id="fakead3p_0">fake3pAd send render-start w/o width or height</h2>
         data-id='0'
         data-url='https://lh3.googleusercontent.com/pSECrJ82R7-AqeBCOEPGPM9iG9OEIQ_QXcbubWIOdkY=w400-h300-no-n'
         data-valid='true'
+        data-native-io='true'
+        data-unlisten-io='true'
```
Comment (lannka) [inline test/manual/fakead3p.amp.html]: nit, maybe pass in a number as seconds?

File: ads/_ping_.js [hunk 5]
```diff
@@ -22,7 +22,11 @@ import {dev} from '../src/log';
  */
 export function _ping_(global, data) {
   global.document.getElementById('c').textContent = data.ping;
-
+  if (!data.nativeIntersectionObserver) {
+    const nullIO = () => {};
+    nullIO.prototype = null;
```
Comment (jridgewell) [inline ads/_ping_.js]: This is an error, arrow functions do not have prototypes. Also, setting a prototype to null doesn't prevent instances from getting a prototype chain, only `Object.create(null)` does that.

Comment (zhouyx) [inline ads/_ping_.js]: hmmm, but it did work though. What should I do to assign the `window.IntersectionObserver` to null in this case. type-check require it to be a function, and this is my hack to fake native IntersectionObserver 😢

Comment (jridgewell) [inline ads/_ping_.js]: Use a function declaration instead.

```js
function nullIO() {};
nullIO.prototype = Object.create(null); // Note you'll have to update your prototype check
```

Comment (zhouyx) [inline ads/_ping_.js]: #6527

**Pull request merged**